### ISSUE TITLE
terraform-ls not working

### ISSUE THREAD

**Issue by pbnj** [yegappan/lsp#189]
Error message is shown: `LSP server exited with status 0`

Installed `terraform-ls` via `brew install terraform-ls`

LSP vim plugin installed via `git clone https://github.com/yegappan/lsp ~/.vim/pack/pbnj/start/lsp`

Relevant Vim configuration:

```vim
" in ~/.vim/vimrc
" lsp
let lspServers = [
      \   #{
      \     filetype: ['vim'],
      \     path: '/opt/homebrew/bin/vim-language-server',
      \     args: ['--stdio'],
      \   },
      \   #{
      \     filetype: ['terraform', 'tf'],
      \     path: '/opt/homebrew/bin/terraform-ls',
      \     args: ['serve'],
      \   },
      \ ]

call LspAddServer(lspServers)

" also tried...
" augroup lsp
"   autocmd!
"   autocmd VimEnter * call LspAddServer(lspServers)
" augroup END
```

Result of `:LspShowAllServers`

```text
Filetype Information
====================
Filetype: 'vim'
Server Path: '/opt/homebrew/bin/vim-language-server'
Status: Not running

Filetype: 'tf'
Server Path: '/opt/homebrew/bin/terraform-ls'
Status: Not running

Filetype: 'terraform'
Server Path: '/opt/homebrew/bin/terraform-ls'
Status: Not running

Buffer Information
==================
Buffer: 'data.tf'
Server Path: '/opt/homebrew/bin/terraform-ls'
Status: Not running
```

Confirmed filetype is correct:

```vim
" in data.tf
:verbose set ft?
  filetype=terraform
        Last set from ~/.dotfiles/vim/.vim/pack/pbnj/start/vim-polyglot/autoload/polyglot/init.vim line 1910
```

Confirmed `vim-language-server` works just fine.

**Comment by lithammer** [comment]
Try enabling server debug and then restart the server:

```
:LspServerDebug on
:LspServerRestart
```

And look in the debug log for clues:

```
tail -F /tmp/lsp-server*
```

**Comment by andlrc** [comment]
It seems like `terraform-ls` is quite picky about their expected content-type:

```
2023/03/28 22:32:43 opts.go:205: Server signaled to stop with err=content type mismatch: got "application/vim-jsonrpc; charset=utf-8", want "application/vscode-jsonrpc; charset=utf-8"
2023/03/28 22:32:43 opts.go:205: Error reading from client: content type mismatch: got "application/vim-jsonrpc; charset=utf-8", want "application/vscode-jsonrpc; charset=utf-8"
```

The `terraform-ls` uses a package https://github.com/creachadair/jrpc2 for the jsonrpc part of the code, which AFAI can read requires the clien to send the header: "application/vscode-jsonrpc; charset=utf-8".  This is also what the [specification states should be supported](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#headerPart).

The string `application/vim-jsonrpc; charset=utf-8` is hard coded within vim, so there isn't an easy fix.

I see three potential ways this could be addresses:

1. I don't know anything about golang, so maybe someone might be able to identify that it's possible to provide another header when using https://github.com/creachadair/jrpc2, and if so then it might be worth opening an issue at https://github.com/hashicorp/terraform-ls
2. It might be worth to open an issue at https://github.com/hashicorp/terraform-ls and https://github.com/creachadair/jrpc2 to request the user to be able to specify another header send by the client.
3. It could make sense for vim to be able to send a custom header, or at least follow the LSP specifications.

This is my take-away with a 10 minutes debug session, so I might have missed a quite a few details.

**Comment by yegappan** [comment]
When I added the support for the LSP mode to Vim (https://github.com/vim/vim/commit/9247a221ce7800c0ae1b3487112d314b8ab79f53), I used vim-jsonrpc, because the LSP specification doesn't state that the Content-type must be set to "application/vscode-jsonrpc; charset=utf-8".  Most of the other language servers doesn't check for vscode in the content-type.  It might be better to open an issue against the terraform-ls to support other content-type headers.
